[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32242 [flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN**
* #32138 use gtest asserts in ProcessGroupGlooTest instead of other checks
* #32134 [flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest

TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

rohan-varma committed:
[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN

TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN"


TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -347,6 +347,13 @@ void testRecv(const std::string& path) {
+  auto s = std::getenv("PYTORCH_TEST_WITH_TSAN");
+  if (s && strcmp(s, "1") == 0) {
+    return;
+  }
```
Can we have a function in TestUtils.hpp that does this and just call it in both places here?

rohan-varma replied:
Sounds good

rohan-varma committed:
Update on "[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN"


TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 8e6535b2:


* **2/4** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=ef0f96e92faeaa8e5e0aa3fae887003e20da683f&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base ef0f96e9 since Jan 15
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **2/4** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8e6535b2bbc409ee561e210c11ba257bc786e44a).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4264893) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86758096) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86758096))
<details>
<summary>
<code>Jan 16 01:43:43 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/TestUtils.hpp:68:15: error: use of undeclared identifier 'strcmp'</code>
</summary>

```
Jan 16 01:43:40   return s && strcmp(s, "1") == 0; 
Jan 16 01:43:40               ^ 
Jan 16 01:43:40 1 error generated. 
Jan 16 01:43:40 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/HashStoreTest.cpp.o] Error 1 
Jan 16 01:43:40 caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/HashStoreTest.cpp.o' failed 
Jan 16 01:43:40 CMakeFiles/Makefile2:11712: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/all' failed 
Jan 16 01:43:40 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/all] Error 2 
Jan 16 01:43:40 make[1]: *** Waiting for unfinished jobs.... 
Jan 16 01:43:43 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/test/TCPStoreTest.cpp:1: 
Jan 16 01:43:43 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/StoreTestCommon.hpp:4: 
Jan 16 01:43:43 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/TestUtils.hpp:68:15: error: use of undeclared identifier 'strcmp' 
Jan 16 01:43:43   return s && strcmp(s, "1") == 0; 
Jan 16 01:43:43               ^ 
Jan 16 01:43:43 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/test/TCPStoreTest.cpp:10: 
Jan 16 01:43:43 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/TCPStore.hpp:8: 
Jan 16 01:43:43 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/Utils.hpp:225:21: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Jan 16 01:43:43   for (int i = 1; i < tensors.size(); ++i) { 
Jan 16 01:43:43                   ~ ^ ~~~~~~~~~~~~~~ 
Jan 16 01:43:43 1 warning and 1 error generated. 
Jan 16 01:43:43 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/TCPStoreTest.dir/TCPStoreTest.cpp.o] Error 1 
Jan 16 01:43:43 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/TCPStoreTest.dir/all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4264887) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86758269) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86758269))
<details>
<summary>
<code>Jan 16 01:43:52 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/TestUtils.hpp:68:28: error: 'strcmp' was not declared in this scope</code>
</summary>

```
Jan 16 01:43:51    return s && strcmp(s, "1") == 0; 
Jan 16 01:43:51                             ^ 
Jan 16 01:43:51 caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/HashStoreTest.cpp.o' failed 
Jan 16 01:43:51 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/HashStoreTest.cpp.o] Error 1 
Jan 16 01:43:51 CMakeFiles/Makefile2:11359: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/all' failed 
Jan 16 01:43:51 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/HashStoreTest.dir/all] Error 2 
Jan 16 01:43:51 make[1]: *** Waiting for unfinished jobs.... 
Jan 16 01:43:52 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/StoreTestCommon.hpp:4:0, 
Jan 16 01:43:52                  from /var/lib/jenkins/workspace/torch/lib/c10d/test/FileStoreTest.cpp:1: 
Jan 16 01:43:52 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/TestUtils.hpp: In function 'bool c10d::test::isTSANEnabled()': 
Jan 16 01:43:52 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/test/TestUtils.hpp:68:28: error: 'strcmp' was not declared in this scope 
Jan 16 01:43:52    return s && strcmp(s, "1") == 0; 
Jan 16 01:43:52                             ^ 
Jan 16 01:43:52 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/FileStoreTest.dir/FileStoreTest.cpp.o] Error 1 
Jan 16 01:43:52 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/FileStoreTest.dir/all] Error 2 
Jan 16 01:43:52 caffe2/lib_c10d/test/CMakeFiles/FileStoreTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/FileStoreTest.dir/FileStoreTest.cpp.o' failed 
Jan 16 01:43:52 CMakeFiles/Makefile2:11401: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/FileStoreTest.dir/all' failed 
Jan 16 01:43:59 [ 93%] Linking CXX executable ../../../bin/ProcessGroupGlooTest 
Jan 16 01:44:00 [ 93%] Built target ProcessGroupGlooTest 
Jan 16 01:44:00 make: *** [all] Error 2 
Jan 16 01:44:00 Makefile:138: recipe for target 'all' failed 
```

</details>

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_xenial_py3_clang5_asan_build](https://dr.pytorch.org/build-details.html?build_id=86758720) from Jan 15 until Jan 15



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

rohan-varma committed:
Update on "[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN"


TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN"


TSAN and fork don't play well together, so skip this test if we're
building under TSAN. It will still run in other modes.

Differential Revision: [D19416113](https://our.internmc.facebook.com/intern/diff/D19416113/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19416113/)!

[ghstack-poisoned]
